### PR TITLE
[expo-structured-headers] add iOS parsing implementation and tests

### PR DESCRIPTION
**Pull request by esamelson**
# Why

Depends on https://github.com/expo/expo/pull/11842

Support parsing of structured headers (https://httpwg.org/http-extensions/draft-ietf-httpbis-header-structure.html), so expo-updates clients can read such headers from EAS update.

This PR adds an Objective C implementation of the parsing spec from the above link. It also adds generated implementations of the entire [official test suite](https://github.com/httpwg/structured-field-tests). Finally, it adds one extra piece of functionality -- a method to parse ignoring parameters (meaning, parse them but do not include them in the return value). This simplifies the structure of the returned object for callers that don't care about parameters (like expo-updates).

The one piece of the specification that we stray from here is that we ignore the ordering of dictionaries. This lets us use simple `NSDictionary` objects in the return values, and we won't need any ordering for the use case we have in mind.

cc @jkhales 

# How

(1st commit) wrote objective C implementation of the parser, following section 4.2 of the specification. Also used https://github.com/apple/swift-http-structured-headers (Swift) and https://github.com/reschke/structured-fields (Java) as reference implementations.

(2nd commit) Next, I had to add some funky stuff in order to get the official test suite to run without a ton of preprocessing. The official test suite  is a bunch of json files with `raw` and `expected` fields, so I wrote a method to loop through these and run each test automatically. However, the `expected` field is in a slightly different format than what we want to return from the parser:
- binary values are represented by an object of the format `{__type: 'binary', value: 'base32-encoded-string'}`, whereas we return `NSData` objects.
- tokens are represented by an object of the format `{__type: 'token', value: 'string'}` whereas we just return simple `NSString` objects.
- dictionaries are represented by an array of tuples because they are ordered in the specification. Our obj-C implementation ignores the ordering of dictionaries and just returns simple `NSDictionary` objects. This would be the trickiest one to try to preprocess, because it's difficult to know if an array of tuples is meant to be a dictionary or not.

In order to properly account for these inside of objects/arrays that can be nested arbitrarily deep, I swizzled the `isEqual:` methods for `NSDictionary` and `NSArray` to make them return true in the above 3 cases, if the values they're representing are equivalent. This should not affect anything outside of the tests.

(3rd commit) I also committed the raw test files from https://github.com/httpwg/structured-field-tests as well as a script that automatically copies them into the respective .h and .m files, so they're easy to update in the future if we want to.

(4th commit) Finally, I added a few more simple tests to ensure the `ignoringParameters` option works as expected. All the examples come from the official test suite.

# Test Plan

All the tests pass 🎉 

# Todos

Need to set up a way to run the XC unit tests in individual packages. Right now, each time I want to run tests, I manually modify the `Exponent` project's podfile to include the Tests subspec, reinstall pods, and run them in the Exponent project. Can commit this if we think it makes sense. cc @ide

### CODE REVIEW

File: packages/expo-structured-headers/ios/EXStructuredHeaders/EXStructuredHeadersParser.m [hunk 1]
```diff
@@ -0,0 +1,534 @@
...
+
+  if (![self compareNextChar:'"']) {
+    if (error) *error = [self errorWithMessage:@"String must begin with the character '\"'"];
+    return nil;
+  }
+
+  [self advance];
+  while ([self hasRemaining]) {
+    unichar nextChar = [self consume];
+    if (nextChar == '\\') {
```
Comment (jkhales) [inline packages/expo-structured-headers/ios/EXStructuredHeaders/EXStructuredHeadersParser.m:335]: isn't this 2 characters?

Comment (jkhales) [reply packages/expo-structured-headers/ios/EXStructuredHeaders/EXStructuredHeadersParser.m:335]: Or is this equivalent to `\u005c` in obj c?

**Review by jkhales — APPROVED** [review]
Thanks for this mammoth PR.

Left a few comments.

File: packages/expo-structured-headers/ios/EXStructuredHeaders/EXStructuredHeadersParser.m (hunk 1 above)
Comment (ide) [reply packages/expo-structured-headers/ios/EXStructuredHeaders/EXStructuredHeadersParser.m:335]: It's a single character -- it's a unichar, not a string.

File: packages/expo-structured-headers/ios/EXStructuredHeaders/EXStructuredHeadersParser.m [hunk 2]
```diff
@@ -0,0 +1,534 @@
...
+    _fieldType = fieldType;
+    _shouldIgnoreParameters = shouldIgnoreParameters;
+  }
+  return self;
+}
+
+- (nullable id)parseStructuredFieldsWithError:(NSError ** _Nullable)error
+{
+  // 4.2
+  // TODO: ASCII
```
[the remark this answers is not in the document]
Comment (esamelson) [reply packages/expo-structured-headers/ios/EXStructuredHeaders/EXStructuredHeadersParser.m]: thanks 🤦 added this check here!

**Pull request merged**